File: p1/main.cpp
class Foo {
    private:
        std::mutex mtx;
        std::condition_variable cv;
        int count = 0;        
        void printNth(std::function<void()> printFunction, int expectedCount) {
            {
                std::unique_lock<std::mutex> lock(mtx);
                cv.wait(lock, [this, expectedCount]{ return (count == expectedCount) && !isPrinting; });
            }
            printFunction();
            {
                std::unique_lock<std::mutex> lock(mtx);
                count = expectedCount + 1;
                cv.notify_all();
            }
        }

    public: 
        void first(std::function<void()> printFirst) {
            printNth(printFirst, 0);
        }

        void second(std::function<void()> printSecond) {
            printNth(printSecond, 1);
        }

        void third(std::function<void()> printThird) {
            printNth(printThird, 2);
        }
};

File: p1/main.java
class Foo {
    volatile int count;
    Object o;

    public Foo() {
        count = 1;
        o = new Object();
    }

    void printNth(Runnable printFunction, int expectedCount) throws InterruptedException {
        synchronized(this){
            while(count != expectedCount){
                wait();
            }
        }
        printFunction.run();
        synchronized(this){
            count = expectedCount + 1;
            notifyAll();
        }
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printNth(printFirst, 1);
    }

    public void second(Runnable printSecond) throws InterruptedException {
        printNth(printSecond, 2);
    }

    public void third(Runnable printThird) throws InterruptedException {
        printNth(printThird, 3);
    }
}
